File: server/hermes.py
"""Hermes Einrichtungs-Service tracking with corrected status semantics."""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any
from urllib.parse import quote

HERMES_API = "https://myhes.de/api/request/auftragsdaten"
MAX_HERMES_RESPONSE_BYTES = 2_000_000
DEFAULT_TIMEOUT = 15


def status_for_id(raw_status_id: object) -> str:
    """Map HES' ordered status bands without treating exceptions as delivered."""

    if not isinstance(raw_status_id, (str, bytes, bytearray, int, float)):
        return "pending"
    try:
        status_id = int(raw_status_id)
    except (TypeError, ValueError):
        return "pending"
    if status_id >= 50_000:
        return "exception"
    if status_id >= 40_000:
        return "delivered"
    if status_id >= 30_000:
        return "out_for_delivery"
    if status_id >= 20_000:
        return "in_transit"
    return "pending"


def event_stage_for_id(raw_status_id: object) -> str:
    status = status_for_id(raw_status_id)
    return "failed_attempt" if status == "exception" else status


def parse_tracking_response(payload: object) -> dict[str, Any]:
    if not isinstance(payload, dict):
        raise ValueError("Hermes returned an invalid tracking response")
    body = payload.get("body", payload)
    if not isinstance(body, dict):
        raise ValueError("Hermes returned an invalid tracking response")
    order = body.get("auftragsdaten") or {}
    if not isinstance(order, dict):
        raise ValueError("Hermes returned an invalid tracking response")
    journey = order.get("statusjourneyDto") or {}
    if not isinstance(journey, dict):
        raise ValueError("Hermes returned an invalid tracking response")

    raw_events: list[dict[str, Any]] = []
    for key in ("auftragstatusdaten", "statusdaten"):
        values = journey.get(key) or []
        if isinstance(values, list):
            raw_events.extend(event for event in values if isinstance(event, dict))
    raw_events.sort(
        key=lambda event: str(event.get("sendungsstatusBuchungszeitpunkt") or ""),
        reverse=True,
    )
    events = [
        {
            "time": str(event.get("sendungsstatusBuchungszeitpunkt") or ""),
            "location": "",
            "description": str(event.get("sendungsstatus") or "Tracking update"),
            "stage": event_stage_for_id(event.get("sendungsstatusId")),
        }
        for event in raw_events
    ]
    status = status_for_id(raw_events[0].get("sendungsstatusId")) if raw_events else "pending"
    return {
        "status": status,
        "last_status_text": events[0]["description"] if events else "",
        "last_update": events[0]["time"] if events else None,
        "expected_delivery": order.get("lieferdatum") or order.get("hesBasicLieferterminZeit"),
        "timezone": "Europe/Zurich",
        "events": events,
    }


class HermesTracker:
    def __init__(self, timeout: int = DEFAULT_TIMEOUT) -> None:
        self.timeout = timeout

    def fetch(self, tracking_number: str) -> dict[str, Any]:
        request = urllib.request.Request(
            f"{HERMES_API}?parcelNumber={quote(tracking_number)}",
            headers={
                "Accept": "application/json",
                "User-Agent": "Mozilla/5.0 (compatible; SwissDeliveryTracker/1.0)",
            },
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                raw = response.read(MAX_HERMES_RESPONSE_BYTES + 1)
        except (urllib.error.HTTPError, urllib.error.URLError, TimeoutError) as exc:
            raise RuntimeError("Hermes tracking is unavailable") from exc
        if len(raw) > MAX_HERMES_RESPONSE_BYTES:
            raise RuntimeError("Hermes returned an unexpectedly large response")
        try:
            payload = json.loads(raw)
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise ValueError("Hermes returned an invalid tracking response") from exc
        return parse_tracking_response(payload)

File: src/server/hermes.ts
import { fetchBounded, parseJsonBytes } from './boundedFetch';
import type { CarrierResult, CarrierStatus } from './carrierResult';
import { isRecord, type JsonObject } from './types';

const HERMES_API = 'https://myhes.de/api/request/auftragsdaten';
const HERMES_STATUS = new Map<number, CarrierStatus>([
  [40, 'pending'],
  [100, 'in_transit'],
  [190, 'in_transit'],
  [300, 'in_transit'],
  [307, 'in_transit'],
  [318, 'exception'],
  [319, 'exception'],
  [320, 'exception'],
  [321, 'exception'],
  [314, 'in_transit'],
  [315, 'in_transit'],
  [500, 'in_transit'],
  [701, 'delivered'],
  [702, 'delivered'],
  [700, 'delivered'],
  [720, 'delivered'],
  [721, 'delivered'],
  [722, 'delivered'],
  [728, 'delivered'],
  [731, 'delivered'],
  [740, 'delivered'],
  [742, 'delivered'],
  [430, 'out_for_delivery'],
]);

export class HermesTrackingError extends Error {
  readonly status = 404;

  constructor() {
    super('Hermes could not locate the shipment');
    this.name = 'HermesTrackingError';
  }
}

function normalizedText(raw: unknown): string {
  return String(raw ?? '').toLocaleLowerCase('de-DE').normalize('NFKD').replace(/\p{M}/gu, '');
}

export function hermesStatus(rawStatusId: unknown, rawDescription: unknown = ''): CarrierStatus {
  if (!['string', 'number'].includes(typeof rawStatusId)) return 'pending';
  const statusId = Number(rawStatusId);
  if (!Number.isFinite(statusId)) return 'pending';
  const known = HERMES_STATUS.get(statusId);
  if (known) return known;
  const description = normalizedText(rawDescription);
  if (/(nicht zugestellt|fehlgeschlagen|storniert|retour|zuruck)/.test(description)) {
    return 'exception';
  }
  if (/(erfolgreich zugestellt|ware geliefert|wurde .* zugestellt|delivered)/.test(description)) {
    return 'delivered';
  }
  if (/(befindet sich auf tour|fahrzeugbeladung|ankunft bei kundenadresse|out for delivery)/
    .test(description)) return 'out_for_delivery';
  if (description) return statusId === 40 ? 'pending' : 'in_transit';
  return 'pending';
}

function eventStage(rawStatusId: unknown, rawDescription: unknown): string {
  const status = hermesStatus(rawStatusId, rawDescription);
  if (status === 'exception') return 'failed_attempt';
  if (status === 'pending') return 'registered';
  return status;
}

function normalizeHermesTrackingNumber(raw: unknown): string {
  return String(raw ?? '').replace(/[\s.-]/g, '').toUpperCase();
}

export function parseHermesTrackingResponse(
  payload: unknown,
  trackingNumber: string,
): CarrierResult {
  if (!isRecord(payload)) throw new TypeError('Hermes returned an invalid tracking response');
  const body = isRecord(payload.body) ? payload.body : payload;
  if (!isRecord(body.auftragsdaten)) {
    throw new TypeError('Hermes returned an invalid tracking response');
  }
  const order = body.auftragsdaten;
  if (
    !order.lieferscheinnummer
    || normalizeHermesTrackingNumber(order.lieferscheinnummer)
      !== normalizeHermesTrackingNumber(trackingNumber)
  ) throw new TypeError('Hermes returned a different shipment');
  const journey = isRecord(order.statusjourneyDto) ? order.statusjourneyDto : {};
  if (journey.auftragstatusdaten !== undefined && !Array.isArray(journey.auftragstatusdaten)) {
    throw new TypeError('Hermes returned invalid tracking history');
  }
  // Hermes's own public UI treats auftragstatusdaten as the customer-facing
  // timeline. statusdaten contains a second, internal operational stream with
  // duplicate scans and different identifiers, so it is intentionally ignored.
  const rawEvents: JsonObject[] = Array.isArray(journey.auftragstatusdaten)
    ? journey.auftragstatusdaten.filter(isRecord)
    : [];
  // The public API answers a valid, unknown 8- or 9-digit consignment with a
  // synthetic order whose identifying/status fields are all null. Do not turn
  // that placeholder into a real-looking pending parcel.
  if (
    rawEvents.length === 0
    && order.auftragId == null
    && order.auftragsart == null
    && order.statusjourneyDto == null
  ) throw new HermesTrackingError();
  const meaningfulEvents = rawEvents.filter((event) => (
    String(event.sendungsstatus ?? '').trim()
    && String(event.sendungsstatusBuchungszeitpunkt ?? '').trim()
  ));
  meaningfulEvents.sort((left, right) => String(right.sendungsstatusBuchungszeitpunkt ?? '')
    .localeCompare(String(left.sendungsstatusBuchungszeitpunkt ?? '')));
  const events = meaningfulEvents.map((event) => ({
    time: String(event.sendungsstatusBuchungszeitpunkt ?? ''),
    location: '',
    description: String(event.sendungsstatus),
    stage: eventStage(event.sendungsstatusId, event.sendungsstatus),
  }));
  return {
    status: meaningfulEvents[0]
      ? hermesStatus(meaningfulEvents[0].sendungsstatusId, meaningfulEvents[0].sendungsstatus)
      : 'pending',
    last_status_text: events[0]?.description ?? '',
    last_update: events[0]?.time || null,
    expected_delivery: typeof order.lieferdatum === 'string'
      ? order.lieferdatum
      : typeof order.hesBasicLieferterminZeit === 'string'
        ? order.hesBasicLieferterminZeit
        : null,
    timezone: 'Europe/Zurich',
    events,
  };
}

export class HermesTracker {
  constructor(readonly timeoutMs = 15_000) {}

  async fetch(trackingNumber: string): Promise<CarrierResult> {
    const url = new URL(HERMES_API);
    url.searchParams.set('parcelNumber', trackingNumber);
    const { bytes } = await fetchBounded(url, {
      headers: {
        Accept: 'application/json',
        'User-Agent': 'Mozilla/5.0 (compatible; SwissDeliveryTracker/1.0)',
      },
    }, {
      provider: 'Hermes tracking',
      timeoutMs: this.timeoutMs,
    });
    return parseHermesTrackingResponse(parseJsonBytes(bytes, 'Hermes'), trackingNumber);
  }
}
